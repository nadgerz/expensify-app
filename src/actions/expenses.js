// ADD_EXPENSE
import uuid from 'uuid';

// use push
// attach then callback
// dispatch action
// redirect

// WHAT WE DO:
// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// components call action generator
// action generator returns function
// component dispatches function (?)
// function runs (has the ability to dispatch other actions and do whatvere it wants)
export const addExpense = ( {
                              description = '',
                              note = '',
                              amount = 0,
                              createdAt = 0
                            } = {} ) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note,
    createdAt,
    amount
  }
});

export const startAddExpense = ( expenseData = {} ) => {
  return ( dispatch ) => {
    const {
            description = '',
            note        = '',
            amount      = 0,
            createdAt   = 0
          } = expenseData;
  };
};

// REMOVE_EXPENSE
export const removeExpense = ( {
                                 id
                               } = {} ) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = ( id, updates ) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
