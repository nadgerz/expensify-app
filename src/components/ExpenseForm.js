import React, { Component } from 'react';

import moment from 'moment';

import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';


export default class ExpenseForm extends Component {
  onDescriptionChange = ( event ) => {
    const description = event.target.value;
    
    this.setState( () => ({ description }) );
  };
  onNoteChange = ( event ) => {
    const note = event.target.value;
    
    this.setState( () => ({ note }) );
  };
  onAmountChange = ( event ) => {
    const amount = event.target.value;
    
    if ( !amount || amount.match( /^\d+(\.\d{0,2})?$/ ) ) {
      this.setState( () => ({ amount }) );
    }
  };
  onDateChange = ( createdAt ) => {
    
    this.setState( () => ({ createdAt }) );
  };
  onDateChange = ( createdAt ) => {
    
    if ( createdAt ) {
      this.setState( () => ({ createdAt }) );
    }
  };
  onFocusChange = ( { focused } ) => {
    
    this.setState( () => ({ calendarFocused: focused }) );
  };
  onSubmit = ( event ) => {
    event.preventDefault();
    
    // console.log( event );
    
    if ( !this.state.description || !this.state.amount ) {
      this.setState( () => ({ error: 'Please provide description and amount' }) );
    }
    else {
      this.setState( () => ({ error: '' }) );
      
      this.props.onSubmit( {
                             description: this.state.description,
                             amount: parseFloat( this.state.amount, 10 ) * 100,
                             createdAt: this.state.createdAt.valueOf(),
                             note: this.state.note
                           } );
    }
  };
  
  constructor( props ) {
    super( props );
    
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: '',
      createdAt: moment(),
      calendarFocused: false,
      error: ''
    };
    
  }
  
  render() {
    return (
      <div>
        
        {
          this.state.error && <p>ERROR: {this.state.error}</p>
        }
        
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            value={this.state.description}
            autoFocus
            onChange={this.onDescriptionChange}
          />
          
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          
          <textarea
            placeholder={'Add a note for your expense (optional)'}
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          
          <button>Add Expense</button>
        </form>
      
      </div>
    );
  }
}
