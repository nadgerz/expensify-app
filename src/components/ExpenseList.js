import React from 'react';

import { connect } from 'react-redux';


const ExpenseList = () => (
  <div>
    <h1>Expense List</h1>
  </div>
);

const ConnectedExpenseList = connect()( ExpenseList );

export default ExpenseList;