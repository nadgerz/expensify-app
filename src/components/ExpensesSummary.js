import React from 'react';

import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';


export const ExpensesSummary = ( props ) => {
  
  const expensStr = props.expenses.length === 1 ? 'expense' : 'expenses';
  console.log( props.expenses );
  
  return (
    <div>
      {expensStr}
    </div>
  );
};

const mapStateToProps = ( state ) => {
  return {
    expenses: selectExpenses( state.expenses, state.filters )
  };
};

export default connect( mapStateToProps )( ExpensesSummary );
