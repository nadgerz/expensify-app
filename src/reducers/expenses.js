// Expenses Reducer

const expensesReducerDefaultState = [];

export default ( state = expensesReducerDefaultState, action ) => {
  
  switch ( action.type ) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    
    case 'REMOVE_EXPENSE':
      return state.filter( ( { id } ) => id !== action.id );
    
    case 'EDIT_EXPENSE':
      // console.log( 'hmm' );
      // console.log( action.id );
      // console.log( action.updates );
      
      return state.map( ( expense ) => {
                          // console.log( `EXPENSE AMOUNT: ${expense.amount}` );
        
                          if ( expense.id === action.id ) {
                            return {
                              ...expense,
                              ...action.updates
                            };
                          }
                          else {
                            return expense;
                          }
                        }
      );
    
    default:
      // console.log( `WARNING: Unknown action type: ${action.type}` );
      return state;
  }
};

