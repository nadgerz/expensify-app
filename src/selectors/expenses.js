// Get visible expenses
export default ( expenses, { text, sortBy, startDate, endDate } ) => {
  return expenses.filter( ( expense ) => {
    const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    
    const textMatch = expense.description.toLocaleLowerCase().includes( text.toLocaleLowerCase() );
    console.log( text.toLocaleLowerCase() );
    console.log( textMatch );
    
    return startDateMatch && endDateMatch && textMatch;
  } );
};
