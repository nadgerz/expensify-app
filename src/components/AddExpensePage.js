import React from 'react';
import { connect } from 'react-redux';

import { addExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';


const AddExpensePage = ( props ) => (
  <div>
    <h1>Add Expense</h1>
    
    <ExpenseForm
      onSubmit={( expense ) => {
        // console.log( 'ADD!' );
        // console.log( expense );
        
        // props.dispatch( addExpense( expense ) );
        props.onSubmit( expense );
        
        props.history.push( '/' );
      }}
    />
  </div>
);

const mapDispatchToProps = ( dispatch ) => {
  return {
    onSubmit: ( expense ) => dispatch( addExpense( expense ) )
  };
};

//
//    Do not need anything from state, so we can have an empty connect()
//
export default connect( undefined, mapDispatchToProps )( AddExpensePage );