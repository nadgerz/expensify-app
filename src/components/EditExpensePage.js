import React from 'react';
import { connect } from 'react-redux';

import { editExpense, removeExpense } from '../actions/expenses';

import ExpenseForm from './ExpenseForm';


const EditExpensePage = ( props ) => {
  console.log( props );
  
  const handleOnClick = ( id ) => {
    props.dispatch( removeExpense( { id } ) );
    
    props.history.push( '/' );
  };
  
  return (
    <div>
      <p>Edit Expense</p>
      
      <ExpenseForm
        expense={props.expense}
        
        onSubmit={( expense ) => {
          // console.log( 'EDIT!' );
          // console.log( expense );
          // console.log( props.match.params.id );
          props.dispatch( editExpense( props.expense.id, expense ) );
          
          props.history.push( '/' );
        }}
      />
      
      <button onClick={handleOnClick( props.match.params.id )}>Remove</button>
    
    </div>
  );
};

const mapStateToProps = ( state, props ) => {
  return {
    expense: state.expenses.find( ( expense ) => {
      return expense.id === props.match.params.id;
    } )
  };
};

export default connect( mapStateToProps )( EditExpensePage );