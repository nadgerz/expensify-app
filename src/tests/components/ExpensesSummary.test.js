import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';


test( 'should render ExpensesSummary correctly', () => {
  const wrapper = shallow( <ExpensesSummary expenses={expenses}/> );
  
  expect( wrapper ).toMatchSnapshot();
} );

test( 'should render correct summary for 1 expense', () => {
  const wrapper = shallow( <ExpensesSummary expenses={[expenses[1]]}/> );
  
  expect( wrapper ).toMatchSnapshot();
} );

test( 'should render correct summary for no expense', () => {
  const wrapper = shallow( <ExpensesSummary expenses={[]}/> );
  
  expect( wrapper ).toMatchSnapshot();
} );

