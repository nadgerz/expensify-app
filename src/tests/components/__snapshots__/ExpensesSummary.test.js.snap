// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly 1`] = `
<div>
  expenses
</div>
`;

exports[`should render correct summary for 1 expense 1`] = `
<div>
  expenses
</div>
`;
